Content.Server/Radiation/Systems/RadiationSystem.GridCast.cs

```diff
@@ -0,0 +1,277 @@
+using System.Linq;
+using Content.Server.Radiation.Components;
+using Content.Shared.Physics;
+using Content.Shared.Radiation.Components;
+using Content.Shared.Radiation.Events;
+using Content.Shared.Radiation.Systems;
+using Robust.Shared.Map;
+using Robust.Shared.Physics;
+using Robust.Shared.Timing;
+
+namespace Content.Server.Radiation.Systems;
+
+// main algorithm that fire radiation rays to target
+public partial class RadiationSystem
+{
+    private void UpdateGridcast()
+    {
+        // should we save debug information into rays?
+        // if there is no debug sessions connected - just ignore it
+        var saveVisitedTiles = _debugSessions.Count > 0;
+
+        var stopwatch = new Stopwatch();
+        stopwatch.Start();
+
+        var sources = EntityQuery<RadiationSourceComponent, TransformComponent>().ToArray();
+        var destinations = EntityQuery<RadiationReceiverComponent, TransformComponent>().ToArray();
+        var blockerQuery = GetEntityQuery<RadiationBlockerComponent>();
+        var resistanceQuery = GetEntityQuery<RadiationGridResistanceComponent>();
+        var transformQuery = GetEntityQuery<TransformComponent>();
+
+        // trace all rays from rad source to rad receivers
+        var rays = new List<RadiationRay>();
+        var receivedRads = new List<(RadiationReceiverComponent, float)>();
```

Comment:
Does ValueList work here?
Also you could probably store the entityuid + float instead of the component itself while here.

Reply:
Sure, but why? I would use try comp a few lines latter and in gridcast no components will be deleted

Reply:
Storing value types in a list is most likely faster than storing refs and then getting the refs again later.

Reply:
Let me do benchmark and check. I can't believe that store object reference can be faster than try comp

Reply:
Here is a benchmark that I did. I tested what is faster: store component itself or store entity uid and after that get component.

Here is benchmark for storing reference:
```csharp
var triesCount = 10000;
var refWatch = new Stopwatch();
refWatch.Start();

for (var i = 0; i < triesCount; i++)
{
    var query = EntityQuery<RadiationReceiverComponent>();
    var refsList = new ValueList<RadiationReceiverComponent>();
    foreach (var comp in query)
    {
        refsList.Add(comp);
    }

    foreach (var comp in refsList)
    {
        comp.CurrentRadiation = 100;
    }
}
Logger.Debug($"Ref list: {refWatch.Elapsed.TotalMilliseconds/triesCount}ms");
```

Storing uid:
```csharp
var uidWatch = new Stopwatch();
uidWatch.Start();

for (var i = 0; i < triesCount; i++)
{
    var getQuery = GetEntityQuery<RadiationReceiverComponent>();
    var query = EntityQuery<RadiationReceiverComponent>();
    var uidList = new ValueList<EntityUid>();
    foreach (var comp in query)
    {
        uidList.Add(comp.Owner);
    }

    foreach (var uid in uidList)
    {
        var comp = getQuery.GetComponent(uid);
        comp.CurrentRadiation = 100;
    }
}
Logger.Debug($"Uid list: {uidWatch.Elapsed.TotalMilliseconds/triesCount}ms");
```

Reply:
Here are results that I got in release mode, around 200 entities:
```
[DEBG] root: Ref list: 0.00966524ms
[DEBG] root: Uid list: 0.00921349ms
[DEBG] root: Ref list: 0.00965037ms
[DEBG] root: Uid list: 0.00926044ms
[DEBG] root: Ref list: 0.00947725ms
[DEBG] root: Uid list: 0.00919981ms
[DEBG] root: Ref list: 0.00958079ms
[DEBG] root: Uid list: 0.00932585ms
[DEBG] root: Ref list: 0.00961159ms
[DEBG] root: Uid list: 0.00897007ms
[DEBG] root: Ref list: 0.00921383ms
[DEBG] root: Uid list: 0.00936675ms
[DEBG] root: Ref list: 0.00933282ms
[DEBG] root: Uid list: 0.00919827ms
[DEBG] root: Ref list: 0.0094339ms
[DEBG] root: Uid list: 0.00917139ms
```
It seems like there are about same magnitude.